pkg/util/test/util.go

```diff
@@ -491,35 +482,98 @@ func validateStorageClusterPods(
 	if _, err := task.DoRetryWithTimeout(t, timeout, interval); err != nil {
 		return err
 	}
+
+	return nil
+}
+
+// Set default Node Affinity rules as Portworx Operator would when deploying StorageCluster
+func defaultPxNodeAffinityRules() *v1.NodeAffinity {
+	nodeAffinity := &v1.NodeAffinity{
+		RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+			NodeSelectorTerms: []v1.NodeSelectorTerm{
+				{
+					MatchExpressions: []v1.NodeSelectorRequirement{
+						{
+							Key:      "px/enabled",
+							Operator: v1.NodeSelectorOpNotIn,
+							Values:   []string{"false"},
+						},
+						{
+							Key:      "node-role.kubernetes.io/master",
+							Operator: v1.NodeSelectorOpDoesNotExist,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return nodeAffinity
+}
+
+func validatePortworxNodes(cluster *corev1.StorageCluster, expectedNodes int) error {
+	conn, err := getSdkConnection(cluster)
+	if err != nil {
+		return nil
+	}
+
+	nodeClient := api.NewOpenStorageNodeClient(conn)
+	nodeEnumerateResp, err := nodeClient.Enumerate(context.Background(), &api.SdkNodeEnumerateRequest{})
+	if err != nil {
+		return err
+	}
+
+	actualNodes := len(nodeEnumerateResp.GetNodeIds())
+	if actualNodes != expectedNodes {
+		return fmt.Errorf("expected nodes: %v. actual nodes: %v", expectedNodes, actualNodes)
+	}
+
+	// TODO: Validate Portworx is started with correct params. Check individual options
+	for _, n := range nodeEnumerateResp.GetNodeIds() {
+		nodeResp, err := nodeClient.Inspect(context.Background(), &api.SdkNodeInspectRequest{NodeId: n})
+		if err != nil {
+			return err
+		}
+		if nodeResp.Node.Status != api.Status_STATUS_OK {
+			return fmt.Errorf("node %s is not online. Current: %v", nodeResp.Node.SchedulerNodeName,
+				nodeResp.Node.Status)
+		}
+
+	}
 	return nil
 }
 
-func expectedPods(cluster *corev1.StorageCluster) (int, error) {
+func getExpectedPxNodeNameList(cluster *corev1.StorageCluster) ([]string, error) {
+	var nodeNameListWithPxPods []string
 	nodeList, err := coreops.Instance().GetNodes()
 	if err != nil {
-		return 0, err
+		return nodeNameListWithPxPods, err
 	}
 
 	dummyPod := &v1.Pod{}
 	if cluster.Spec.Placement != nil && cluster.Spec.Placement.NodeAffinity != nil {
 		dummyPod.Spec.Affinity = &v1.Affinity{
 			NodeAffinity: cluster.Spec.Placement.NodeAffinity.DeepCopy(),
 		}
+	} else {
+		dummyPod.Spec.Affinity = &v1.Affinity{
+			NodeAffinity: defaultPxNodeAffinityRules().DeepCopy(),
```

Comment:
nit: No need to deepcopy here. It will return a new object every time anyway.

Reply:
Got it, fixed.